src/libraries/Microsoft.PowerFx.Core/Public/Values/CompileTimeTypeWrapperRecordValue.cs

```diff
@@ -0,0 +1,46 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using Microsoft.PowerFx.Core.Public.Types;
+
+namespace Microsoft.PowerFx.Core.Public.Values
+{
+    // Project the correct compile-time type onto the runtime value. 
+    // Important for union / intersection types, such as Table() or If(). For example:
+    //    First(Table({a:1},{b:2})), result is a record with both fields a and b. 
+    internal class CompileTimeTypeWrapperRecordValue : RecordValue
+    {
+        private readonly RecordValue _inner;
+
+        public static RecordValue AdjustType(RecordType expectedType, RecordValue inner)
+        {
+            // TODO - Could this check be in the IR?  And so have IR add calls to TypeOverrideRecordValue?
```

Comment:
I don't think so, record values with a subset of the expected fields could also be passed in from the host, or from connectors. We wouldn't know that at compile time.